RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc

```diff
@@ -0,0 +1,244 @@
+        jet_tags_[intag.encode()] = consumes<JetTagCollection>(intag);
+      }
+      current.numerator.push_back(intag.encode());
+    }
+
+    for (auto intag : discriminator.getParameter<vInputTag>("denominator")) {
+      if (jet_tags_.find(intag.encode()) == jet_tags_.end()) { // new
+                                                               // probability
+        jet_tags_[intag.encode()] = consumes<JetTagCollection>(intag);
+      }
+      current.denominator.push_back(intag.encode());
+    }
+    discrims_.push_back(current);
+  }
+
+  if (jet_tags_.empty()) {
+    throw cms::Exception("RuntimeError")
+        << "The module BTagProbabilityToDiscriminator is run without any input "
+           "probability to work on!"
+        << std::endl;
+  }
+}
+
+void BTagProbabilityToDiscriminator::produce(edm::Event &iEvent,
+                                             const edm::EventSetup &iSetup) {
+  std::unordered_map<std::string, edm::Handle<JetTagCollection>>
+      tags; // caches jet tags to avoid repetitions
+  size_t size = 0;
+  bool first = true;
+  for (auto entry : jet_tags_) {
+    edm::Handle<JetTagCollection> tmp;
+    iEvent.getByToken(entry.second, tmp);
+    tags[entry.first] = tmp;
+    if (first)
+      size = tmp->size();
+    else {
+      if (tmp->size() != size) {
+        throw cms::Exception("RuntimeError")
+            << "The length of one of the input jet tag collections does not "
+               "match "
+            << "with the others, this is probably due to the probabilities "
+               "belonging to different jet collections, which is forbidden!"
+            << std::endl;
+      }
+    }
+    first = false;
+  }
+
+  // create the output collection
+  // which is a "map" RefToBase<Jet> --> float
+  vector<std::unique_ptr<JetTagCollection>> output_tags;
+  output_tags.reserve(discrims_.size());
+  for (size_t i = 0; i < discrims_.size(); ++i) {
+    output_tags.push_back(std::make_unique<JetTagCollection>(
+        *(tags.begin()->second)) // clone from the first element, will change
+                                 // the content later on
+                          );
+  }
+
+  // loop over jets
+  for (size_t idx = 0; idx < output_tags[0]->size(); idx++) {
+    auto key = output_tags[0]->key(idx); // index access should be OK as well,
+                                         // but a bit less safe, probably WAY
+                                         // faster
+
+    // loop over new discriminators to produce
+    for (size_t disc_idx = 0; disc_idx < output_tags.size(); disc_idx++) {
+      float numerator = 0;
+      for (auto &num : discrims_[disc_idx].numerator)
+        numerator += (*tags[num])[key];
+      float denominator = !discrims_[disc_idx].denominator.empty() ? 0 : 1;
+      for (auto &den : discrims_[disc_idx].denominator)
+        denominator += (*tags[den])[key];
+      float new_value = (denominator != 0 && numerator > 0) ? numerator / denominator : -10.;
```

Comment:
I don't really follow the need to check that numerator>0? Is there an easy explanation?

Reply:
The probability < 0 is used to mark an undefined jet, where there is no information to make a decision (no tracks pass the track pre-selection and no secondary vertices are found). With this module the discriminator defaults to some number, sometimes positive (e.g. the charm tagger which is prob(c)/(prob(b)+prob(c) defaulting at 0.5), hence the need of that guard. I can add a comment.